src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpTraceExporterHelperExtensions.cs

```diff
@@ -43,19 +44,31 @@ public static TracerProviderBuilder AddOtlpExporter(this TracerProviderBuilder b
         /// <param name="configure">Callback action for configuring <see cref="OtlpExporterOptions"/>.</param>
         /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
         public static TracerProviderBuilder AddOtlpExporter(this TracerProviderBuilder builder, Action<OtlpExporterOptions> configure)
-            => AddOtlpExporter(builder, name: null, configure);
+            => AddOtlpExporter(builder, name: null, configure, persistentStorageFactory: null);
+
+        /// <summary>
+        /// Adds OpenTelemetry Protocol (OTLP) exporter to the TracerProvider.
+        /// </summary>
+        /// <param name="builder"><see cref="TracerProviderBuilder"/> builder to use.</param>
+        /// <param name="configure">Callback action for configuring <see cref="OtlpExporterOptions"/>.</param>
+        /// <param name="persistentStorageFactory">Factory for creating an offline storage implementation.</param>
+        /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
+        public static TracerProviderBuilder AddOtlpExporter(this TracerProviderBuilder builder, Action<OtlpExporterOptions> configure, Func<IServiceProvider, PersistentBlobProvider> persistentStorageFactory)
+            => AddOtlpExporter(builder, name: null, configure, persistentStorageFactory);
 
         /// <summary>
         /// Adds OpenTelemetry Protocol (OTLP) exporter to the TracerProvider.
         /// </summary>
         /// <param name="builder"><see cref="TracerProviderBuilder"/> builder to use.</param>
         /// <param name="name">Name which is used when retrieving options.</param>
         /// <param name="configure">Callback action for configuring <see cref="OtlpExporterOptions"/>.</param>
+        /// <param name="persistentStorageFactory">Factory for creating an offline storage implementation.</param>
         /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
         public static TracerProviderBuilder AddOtlpExporter(
             this TracerProviderBuilder builder,
             string name,
-            Action<OtlpExporterOptions> configure)
+            Action<OtlpExporterOptions> configure,
+            Func<IServiceProvider, PersistentBlobProvider> persistentStorageFactory = null)
```

Comment:
We need to remove the default (`= null`) on the parameter here, default parameters are trouble with a lot of overloads.

We could probably make this work, but here's an alternative idea... target options for the extension:

```csharp
public static OtlpExporterOptions ConfigurePersistentStorage(this OtlpExporterOptions options, Func<IServiceProvider, PersistentBlobProvider> persistentStorageFactory)
{
   Guard.ThrowIfNull(options);
   Guard.ThrowIfNull(persistentStorageFactory);

   // If we add PersistentStorageFactory as an internal on options
   options.PersistentStorageFactory = persistentStorageFactory;
   return options;
}
```

That would give us syntax like this:

```chsarp
builder.AddOtlpExporter(
   options => options.ConfigurePersistentStorage(_ => new FileBlobProvider(...)));
```

If you do that, could also have another extension which takes an instance directly:

```csharp
public static OtlpExporterOptions ConfigurePersistentStorage(this OtlpExporterOptions options, PersistentBlobProvider persistentBlobProvider)
{
   Guard.ThrowIfNull(options);
   Guard.ThrowIfNull(persistentBlobProvider);

   options.PersistentStorageFactory = _ => persistentBlobProvider;
   return options;
}
```

Reply:
I'll expand on the above a bit...

I think ideally in a perfect world we would have clean separation between options (settings) and services (extension points). The best way to do that would maybe be to return a builder like...

```csharp
public static OtlpTraceExporterBuilder AddOtlpExporter(this TracerProviderBuilder builder)
{
   return new OtlpTraceExporterBuilder(builder);
}
```

That would allow us to write extensions for OtlpExporterBuilder and have syntax like this...

```csharp
builder.AddOtlpExporter()
   .ConfigureOptions(options => options.Endpoint = new Uri(...))
   .ConfigurePersistentStorage(new FileBlobProvider(...))
   .ConfigureBatchProcessor(...);
```

It is just a question of do we want to try and retrofit that or not 😄 The above extensions on options are a bit of hack to avoid the builder.